### PR TITLE
SM64: Move Randomizer Content Update

### PR DESCRIPTION
**Pull request by Magnemania**
## What is this fixing or adding?

- This adds new options (which default to false) that allow the player to lock up to 10 of Mario's movement abilities behind collectible items.
- This replaces most of the existing logic with a new logic system that parses rules from a basic language (to allow these rules to be easily read by players and updated in the future).  The generated lambdas are simplified if the player does not randomize the moves.
- Regions have been reworked to split levels into several subregions for move randomizer logic.
- Star requirement options have been altered to use percentages, rather than hard numbers, so that the total star count option can be respected.  This ensures that a player will always get the total star count they requested as long as enough locations exist in the world to fill it (whereas the old behavior would subtract stars from the player's total if they select certain other options).  If the player leaves these options at their default values, star requirements will exactly match those in the vanilla game.

## How was this tested?

The Archipelago community has been openly testing with the distributed APWorld since March of this year, and logic improvements have been implemented in response to community feedback.

## If this makes graphical changes, please attach screenshots.

image

### CODE REVIEW

File: worlds/sm64ex/Options.py [hunk 1]
```diff
@@ -96,25 +106,43 @@ class CompletionType(Choice):
...
 
+class StrictMoveRequirements(DefaultOnToggle):
+    """If disabled, Stars that expect certain moves may have to be acquired without them. Only makes a difference
+    if Move Randomization is enabled"""
+    display_name = "Strict Move Requirements"
+
+def getMoveRandomizerOption(action: str):
+    class MoveRandomizerOption(Toggle):
+        """Mario is unable to perform this action until a corresponding item is picked up.
+        If Mario 64 is built with an alternative "nomoverando" branch, this option will not affect Mario."""
```
Comment (N00byKing) [inline worlds/sm64ex/Options.py:120]: I'd rather this say something like `This option is incompatible with builds using a 'nomoverando' branch`.
The `nomoverando` branch prints huge warnings on screen if they attempt to load a game where move rando is enabled, its not like its a cheat code to have it not affect mario (Except if they can ignore that I guess).

File: worlds/sm64ex/Regions.py [hunk 2]
```diff
@@ -5,8 +5,9 @@
     locHMC_table, locLLL_table, locSSL_table, locDDD_table, locSL_table, \
     locWDW_table, locTTM_table, locTHI_table, locTTC_table, locRR_table, \
     locPSS_table, locSA_table, locBitDW_table, locTotWC_table, locCotMC_table, \
-    locVCutM_table, locBitFS_table, locWMotR_table, locBitS_table, locSS_table    
+    locVCutM_table, locBitFS_table, locWMotR_table, locBitS_table, locSS_table
 
+# List of all courses, including secrets, without BitS as that one is static
```
Comment (N00byKing) [inline worlds/sm64ex/Regions.py:10]: Comment is at a weird place. Currently looks like it is just referencing `sm64_level_to_paintings`, which does not include secrets.
In the first place these arent lists, but dicts from the game-specific entrance format to the name of the entrance.
Does make me realize I made the same mistake just below for sm64secrets though ^^'

File: worlds/sm64ex/Rules.py [hunk 3]
```diff
@@ -1,33 +1,54 @@
...
+    if entrance_map[swapdict[entrance]] == destination: # Unlucky :C
         rand = world.random.choice(swapdict.keys())
-        entrance_ids[entrance], entrance_ids[swapdict[rand]] = rand, entrance_ids[entrance]
-        swapdict[rand] = entrance_ids[entrance]
+        entrance_map[entrance], entrance_map[swapdict[rand]] = rand, entrance_map[entrance]
+        swapdict[rand] = entrance_map[entrance]
         swapdict.pop(entrance)
 
-def set_rules(world, player: int, area_connections: dict):
+def set_rules(world, player: int, area_connections, star_costs, move_rando_bitvec: int):
```
Comment (N00byKing) [inline worlds/sm64ex/Rules.py:23]: Why did you remove the typing annotation?

File: worlds/sm64ex/__init__.py [hunk 4]
```diff
@@ -71,17 +102,8 @@ def create_item(self, name: str) -> Item:
...
-    def create_items(self):
-        starcount = self.multiworld.AmountOfStars[self.player].value
-        if (not self.multiworld.EnableCoinStars[self.player].value):
-            starcount = max(35,self.multiworld.AmountOfStars[self.player].value-15)
-        starcount = max(starcount, self.multiworld.FirstBowserStarDoorCost[self.player].value,
-                        self.multiworld.BasementStarDoorCost[self.player].value, self.multiworld.SecondFloorStarDoorCost[self.player].value,
-                        self.multiworld.MIPS1Cost[self.player].value, self.multiworld.MIPS2Cost[self.player].value,
-                        self.multiworld.StarsToFinish[self.player].value)
-        self.multiworld.itempool += [self.create_item("Power Star") for i in range(0,starcount)]
-        self.multiworld.itempool += [self.create_item("1Up Mushroom") for i in range(starcount,120 - (15 if not self.multiworld.EnableCoinStars[self.player].value else 0))]
```
Comment (N00byKing) [inline worlds/sm64ex/__init__.py:83]: Items must be created in `create_items`, per #1460

File: worlds/sm64ex/Rules.py [hunk 5]
```diff
@@ -1,33 +1,54 @@
...
+    randomized_entrances = {**randomized_level_to_paintings, **randomized_level_to_secrets}
     if world.AreaRandomizer[player].value == 3:  # Randomize Courses and Secrets in one pool
-        randomized_entrances = shuffle_dict_keys(world,randomized_entrances)
+        randomized_entrances = shuffle_dict_keys(world, randomized_entrances)
         # Guarantee first entrance is a course
-        swapdict = { entrance: level for (level,entrance) in randomized_entrances }
-        if randomized_entrances[91] not in sm64_paintings_to_level.keys(): # Unlucky :C (91 -> BoB Entrance)
-            rand = world.random.choice(sm64_paintings_to_level.values())
+        swapdict = {entrance: level for (level, entrance) in randomized_entrances.items()}
+        valid_first_courses = valid_move_randomizer_start_courses if move_rando_bitvec > 0 else list(sm64_paintings_to_level.keys())
```
Comment (N00byKing) [inline worlds/sm64ex/Rules.py:49]: Is the cast to list necessary here?

File: worlds/sm64ex/Rules.py [hunk 6]
```diff
@@ -1,33 +1,54 @@
...
+    valid_move_randomizer_start_courses = [
+        "Bob-omb Battlefield", "Jolly Roger Bay", "Cool, Cool, Mountain",
+        "Big Boo's Haunt", "Lethal Lava Land", "Shifting Sand Land",
+        "Dire, Dire Docks", "Snowman's Land"
+    ]  # Excluding WF, HMC, WDW, TTM, THI, TTC, and RR
+    if world.AreaRandomizer[player].value >= 1:  # Some randomization is happening, randomize Courses
         randomized_level_to_paintings = shuffle_dict_keys(world,sm64_level_to_paintings)
+        if world.AreaRandomizer[player].value < 3 and move_rando_bitvec > 0:
+            first_course = world.random.choice(valid_move_randomizer_start_courses)
+            for entrance, painting in randomized_level_to_paintings.items():
```
Comment (N00byKing) [inline worlds/sm64ex/Rules.py:35]: I dont like this loop.
Equivalent but to me nicer looking: `next(key for key, value in dd.items() if value == 'value')
`.
Though tbh I found this on stackoverflow

File: worlds/sm64ex/Rules.py [hunk 7]
```diff
@@ -1,33 +1,54 @@
...
+    keys = list(dictionary.keys())
+    values = list(dictionary.values())
     world.random.shuffle(keys)
-    return dict(zip(keys,values))
+    return dict(zip(keys, values))
 
-def fix_reg(entrance_ids, entrance, destination, swapdict, world):
-    if entrance_ids[entrance] == destination: # Unlucky :C
+def fix_reg(entrance_map, entrance, destination, swapdict, world):
+    if entrance_map[swapdict[entrance]] == destination: # Unlucky :C
```
Comment (N00byKing) [inline worlds/sm64ex/Rules.py:17]: Should probably be `entrance_map[sm64_entrances_to_level[entrance]]` right?

**Review by N00byKing — CHANGES_REQUESTED** [review]
There are some things I missed the first time through I guess, but nothing major.
Keep in mind my python is not really good, so I might've made some mistakes in the review

File: worlds/sm64ex/Options.py (hunk 1 above)
Comment (Magnemania) [reply worlds/sm64ex/Options.py]: Done.

https://github.com/ArchipelagoMW/Archipelago/pull/2569/commits/1cab3efe156f91109f031006808fab88bfe0490f

File: worlds/sm64ex/Regions.py (hunk 2 above)
Comment (Magnemania) [reply worlds/sm64ex/Regions.py]: Removed that comment and cleaned up the secrets comment.

File: worlds/sm64ex/Rules.py (hunk 3 above)
Comment (Magnemania) [reply worlds/sm64ex/Rules.py]: Accident.  (https://github.com/ArchipelagoMW/Archipelago/pull/2569/commits/007aa8133f9d39072b77531a4f8b9a3749f2147b)

File: worlds/sm64ex/__init__.py (hunk 4 above)
Comment (Magnemania) [reply worlds/sm64ex/__init__.py:83]: Making Starcraft 2 comply with that PR is going to be involve rewriting the whole codebase from start-to-finish.  A lot easier to fix it in SM64.  Done.  (https://github.com/ArchipelagoMW/Archipelago/pull/2569/commits/163a61c6f62ec64009df4af4749dc036b3b54773)

File: worlds/sm64ex/Rules.py (hunk 6 above)
Comment (Magnemania) [reply worlds/sm64ex/Rules.py]: I normally use comprehensions everywhere, I'm not sure why I didn't do so here.  Changed.  (https://github.com/ArchipelagoMW/Archipelago/pull/2569/commits/29480bd177069995797952def1dd5390dcb651f0)

File: worlds/sm64ex/Rules.py (hunk 7 above)
Comment (Magnemania) [reply worlds/sm64ex/Rules.py:17]: I'm having some trouble figuring out how swap_dict works, and it doesn't seem to be working correctly before or after that change.  I'm going to have to take a look at it later; might make an enum for Entrances to make it easier to understand.

File: worlds/sm64ex/__init__.py (hunk 4 above)
Comment (NewSoupVi) [reply worlds/sm64ex/__init__.py:83]: Just to be a bit pedantic: It must be done *by* create_items, not *in* create_items - If for some reason you want to make items in generate_early, you can do that

Comment (Magnemania) [reply worlds/sm64ex/__init__.py:83]: > I think you are no longer creating the ExclamationBox items

Why would they not be created?  I haven't done much to change the code in that part.

I cannot create the cannons or exclamation box items in create_items if the options result in them getting hard-placed later, as that would be modifying the item pool after create_items.

File: worlds/sm64ex/Rules.py [hunk 8]
```diff
@@ -1,67 +1,96 @@
...
-            swapdict[rand] = randomized_entrances[91]
-        swapdict.pop("Bob-omb Battlefield")
+        swapdict = {entrance: level for (level, entrance) in randomized_entrances.items()}
+        valid_first_courses = valid_move_randomizer_start_courses if move_rando_bitvec > 0 else list(sm64_paintings_to_level.keys())
+        first_course = randomized_entrances[Entrances.BOB_OMB_BATTLEFIELD]
+        if first_course not in valid_first_courses:  # Unlucky :C
+            rand = world.random.choice(valid_first_courses)
+            rand_entrance = swapdict[rand]
+            randomized_entrances[Entrances.BOB_OMB_BATTLEFIELD], randomized_entrances[rand_entrance] = rand, first_course
+            swapdict.pop(rand)
```
Comment (N00byKing) [inline worlds/sm64ex/Rules.py:55]: Same as above.
Can probably be simplified to just use fix_reg as well.
`fix_reg(randomized_entrances, Entrances.BOB_OMB_BATTLEFIELD, <Courses not in valid_first_courses>, swapdict, world)`
You can then also remove the if cond just above, as its checked in fix_reg

File: worlds/sm64ex/Rules.py [hunk 9]
```diff
@@ -1,67 +1,96 @@
...
+    return dict(zip(keys, values))
 
-def fix_reg(entrance_ids, entrance, destination, swapdict, world):
-    if entrance_ids[entrance] == destination: # Unlucky :C
-        rand = world.random.choice(swapdict.keys())
-        entrance_ids[entrance], entrance_ids[swapdict[rand]] = rand, entrance_ids[entrance]
-        swapdict[rand] = entrance_ids[entrance]
-        swapdict.pop(entrance)
+def fix_reg(entrance_map: Dict[Entrances, str], entrance: Entrances, destination: str,
+            swapdict: Dict[str, Entrances], world, invalid_regions: Set[str] = {}):
```
Comment (N00byKing) [inline worlds/sm64ex/Rules.py:17]: If you are introducing `invalid_regions`, no need to keep `destination` parameter; just use that instead. (Put it as second parameter tho, not at the end)

File: worlds/sm64ex/Rules.py [hunk 10]
```diff
@@ -1,67 +1,96 @@
...
+    if entrance_map[entrance] == destination: # Unlucky :C
+        replacement_regions = [(rand_region, rand_entrance) for rand_region, rand_entrance in swapdict.items()
+                               if rand_region not in invalid_regions and rand_region != destination]
+        rand_region, rand_entrance = world.random.choice(replacement_regions)
+        entrance_map[entrance], entrance_map[rand_entrance] = rand_region, destination
+        swapdict[rand_region] = entrance
+        swapdict.pop(destination)
```
Comment (N00byKing) [inline worlds/sm64ex/Rules.py:24]: Thinking about it, this needs to always pop. Otherwise, on subsequent calls an invalid region might get swapped back in (So always pop entrance_map[entrance])
So for any call to fix_reg, after the call the destination for this entrance will no longer change

File: worlds/sm64ex/Rules.py [hunk 11]
```diff
@@ -1,67 +1,96 @@
...
         # Guarantee COTMC is not mapped to HMC, cuz thats impossible
-        fix_reg(randomized_entrances, "Cavern of the Metal Cap", "Hazy Maze Cave", swapdict, world)
+        fix_reg(randomized_entrances, Entrances.CAVERN_OF_THE_METAL_CAP, "Hazy Maze Cave", swapdict, world)
         # Guarantee BITFS is not mapped to DDD
-        fix_reg(randomized_entrances, "Bowser in the Fire Sea", "Dire, Dire Docks", swapdict, world)
-        if randomized_entrances[191] == "Hazy Maze Cave": # If BITFS is mapped to HMC...
-            fix_reg(randomized_entrances, "Cavern of the Metal Cap", "Dire, Dire Docks", swapdict, world) # ... then dont allow COTMC to be mapped to DDD
+        fix_reg(randomized_entrances, Entrances.BOWSER_IN_THE_FIRE_SEA, "Dire, Dire Docks", swapdict, world)
+        if randomized_entrances[Entrances.BOWSER_IN_THE_FIRE_SEA] == "Hazy Maze Cave": # If BITFS is mapped to HMC...
+            fix_reg(randomized_entrances, Entrances.CAVERN_OF_THE_METAL_CAP, "Dire, Dire Docks", swapdict, world, {"Hazy Maze Cave"}) # ... then dont allow COTMC to be mapped to DDD
```
Comment (N00byKing) [inline worlds/sm64ex/Rules.py:62]: This can be simplified using `invalid_regions`, if you remove the `destination` parameter
```
fix_reg for BitFS
fix_reg for COTMC (invalid regions both HMC and DDD)
```
So only 2 calls to fix_reg

File: worlds/sm64ex/Regions.py [hunk 12]
```diff
@@ -1,189 +1,230 @@
...
-    111: "Wet-Dry World",
-    361: "Tall, Tall Mountain",
-    132: "Tiny-Huge Island (Tiny)",
-    131: "Tiny-Huge Island (Huge)",
-    141: "Tick Tock Clock",
-    151: "Rainbow Ride"
+    locVCutM_table, locBitFS_table, locWMotR_table, locBitS_table, locSS_table
+
+
+class Entrances(int, Enum):
```
Comment (N00byKing) [inline worlds/sm64ex/Regions.py:13]: Not a fan of the name of this enum.
First, `Entrances` could also mean Entrances between regions.
Then, `Entrances` could also mean the string representations of the entrances (`Tiny-Huge Island (Tiny)`).
As a suggestion: The game code refers to the regions as levels (So `LEVEL_THI` for example). Something like `SM64Levels`?

File: worlds/sm64ex/__init__.py (hunk 4 above)
Comment (N00byKing) [reply worlds/sm64ex/__init__.py:83]: My bad, misread

**Review by N00byKing — CHANGES_REQUESTED** [review]
Second round is in

**Comment by Magnemania** [comment]
Addressed the issues in Round 2.  Also fixed an issue with the 1Up generation.

File: worlds/sm64ex/__init__.py [hunk 13]
```diff
@@ -35,22 +35,53 @@ class SM64World(World):
...
+        max_stars = 120
+        if (not self.multiworld.EnableCoinStars[self.player].value):
+            max_stars -= 15
+        self.move_rando_bitvec = 0
+        for action, itemid in action_item_table.items():
+            # HACK: Disable randomization of double jump
+            if action == 'Double Jump': continue
+            if getattr(self.multiworld, f"MoveRandomizer{action.replace(' ','')}")[self.player].value:
+                max_stars -= 1
+                self.multiworld.itempool.append(self.create_item(action))
```
Comment (N00byKing) [inline worlds/sm64ex/__init__.py:60]: Can you move this to `create_items`? Its technically allowed here, but it would be more consistent with the other items.

**Review by N00byKing — COMMENTED** [review]
One last thing, otherwise LGTM!

File: worlds/sm64ex/__init__.py (hunk 13 above)
Comment (Magnemania) [reply worlds/sm64ex/__init__.py]: Missed this comment.  Moved item creation to `create_items` using the same conventions as checking for items in rules.

**Comment by N00byKing** [comment]
I am proud to present: ✅

**Review by N00byKing — APPROVED** [review]

**Comment by N00byKing** [comment]
AP 0.4.4 is out so we should be out of feature freeze. @Magnemania please rebase, then this is hopefully good to go

**Comment by Magnemania** [comment]
> AP 0.4.4 is out so we should be out of feature freeze. @Magnemania please rebase, then this is hopefully good to go

I, uh, hit the Rebase button in Pycharm, and I don't think it rebased correctly.

**Comment by N00byKing** [comment]
RIP Git History... Well, this was going to be squash merged anyway, so we'll do that manually.
- Copy the sm64 world as you have it
- Reset the branch on your end to the AP `main` branch
- Copy it back, commit

So in the end there's only one commit with all changes. As I've already reviewed everything it should be fine.

**Comment by Magnemania** [comment]
> RIP Git History... Well, this was going to be squash merged anyway, so we'll do that manually.
> 
> * Copy the sm64 world as you have it
> * Reset the branch on your end to the AP `main` branch
> * Copy it back, commit
> 
> So in the end there's only one commit with all changes. As I've already reviewed everything it should be fine.

Going to be completely honest with you here, I've been Googling for the past hour and have absolutely no idea how to do that.  I've been using PyCharm exclusively to interface with Git, and it doesn't look like it provides a way for me to reset a remote branch; I didn't think it was even possible to reset a branch.  How do I do that?

**Comment by N00byKing** [comment]
I dont use GUI with git, and have never used pycharm... If you can use a terminal, `git reset --hard origin/main` will reset current branch

File: worlds/sm64ex/Rules.py [hunk 14]
```diff
@@ -40,30 +63,35 @@ def set_rules(world, player: int, area_connections: dict):
             fix_reg(randomized_entrances, SM64Levels.CAVERN_OF_THE_METAL_CAP, {"Hazy Maze Cave"}, swapdict, world)
 
     # Destination Format: LVL | AREA with LVL = LEVEL_x, AREA = Area as used in sm64 code
-    # Cast to int to not rely on availability of SM64Levels enum. Will cause crash in MultiServer otherwise
```
Comment (N00byKing) [inline worlds/sm64ex/Rules.py:43]: You need to restore this comment

Comment (Magnemania) [reply worlds/sm64ex/Rules.py:43]: Nevermind, it looks like I could have undid the force push from the command line, but I did it from PyCharm.  Comment is back.  How do I give you and https://github.com/RBmans credit, again?  He assisted with Rules.py.

Comment (N00byKing) [reply worlds/sm64ex/Rules.py:43]: `Co-authored-by: name <additional-dev-1@example.com>` in the commit message for RBman. You can get the mail from his original commit.

I only did minor things so no need

**Review by N00byKing — APPROVED** [review]
Green Tick number two
